File: src/mode/rt/rtreq/cubes_req.go
package rtreq

import (
	"encoding/json"

	"fmt"

	"github.com/gin-gonic/gin"
	"github.com/t-kawata/mycute/lib/common"
	"github.com/t-kawata/mycute/mode/rt/rtres"
	"github.com/t-kawata/mycute/mode/rt/rtutil"
	"github.com/t-kawata/mycute/model"
	"github.com/t-kawata/mycute/pkg/cuber/types"
	"github.com/t-kawata/mycute/pkg/cuber/validator"
)

type SearchCubesReq struct {
	Name        string `json:"name" binding:"max=50"`
	Description string `json:"description" binding:"max=255"`
	Limit       uint16 `json:"limit" binding:"gte=1,lte=25"`
	Offset      uint16 `json:"offset" binding:"gte=0"`
}

func SearchCubesReqBind(c *gin.Context, u *rtutil.RtUtil) (SearchCubesReq, rtres.SearchCubesRes, bool) {
	ok := true
	req := SearchCubesReq{}
	res := rtres.SearchCubesRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindJSON(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type GetCubeReq struct {
	ID uint `form:"cube_id" binding:"required,gte=1"`
}

func GetCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (GetCubeReq, rtres.GetCubeRes, bool) {
	ok := true
	req := GetCubeReq{ID: common.StrToUint(c.Param("cube_id"))}
	res := rtres.GetCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBind(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type CreateCubeReq struct {
	Name               string `json:"name" binding:"required,max=50"`
	Description        string `json:"description" binding:"max=255"`
	EmbeddingProvider  string `json:"embedding_provider" binding:"required,max=50"`
	EmbeddingModel     string `json:"embedding_model" binding:"required,max=100"`
	EmbeddingDimension uint   `json:"embedding_dimension" binding:"required,gte=1"` // 0は不可
	EmbeddingApiKey    string `json:"embedding_api_key" binding:"required"`
	EmbeddingBaseURL   string `json:"embedding_base_url" binding:"omitempty,max=255"`
}

func CreateCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (CreateCubeReq, rtres.CreateCubeRes, bool) {
	ok := true
	req := CreateCubeReq{}
	res := rtres.CreateCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindJSON(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
		return req, res, ok
	}
	// 埋め込みモデル関連のバリデーション実行
	if err := validator.ValidateEmbeddingConfig(req.EmbeddingProvider, req.EmbeddingModel, req.EmbeddingDimension); err != nil {
		res.Errors = append(res.Errors, rtres.Err{Field: "embedding_provider", Message: fmt.Sprintf("Invalid embedding configuration: %s", err.Error())})
		res.Errors = append(res.Errors, rtres.Err{Field: "embedding_model", Message: fmt.Sprintf("Invalid embedding configuration: %s", err.Error())})
		res.Errors = append(res.Errors, rtres.Err{Field: "embedding_dimension", Message: fmt.Sprintf("Invalid embedding configuration: %s", err.Error())})
		ok = false
	}
	// Live Test Embedding
	if ok {
		// Only run if basic validation passed
		embConfig := types.EmbeddingModelConfig{
			Provider:  req.EmbeddingProvider,
			Model:     req.EmbeddingModel,
			Dimension: req.EmbeddingDimension,
			BaseURL:   req.EmbeddingBaseURL,
			ApiKey:    req.EmbeddingApiKey,
		}

		if err := u.CuberService.VerifyEmbeddingConfiguration(c.Request.Context(), embConfig); err != nil {
			res.Errors = append(res.Errors, rtres.Err{Field: "embedding_config", Message: fmt.Sprintf("Live embedding verification failed: %s", err.Error())})
			ok = false
		}
	}
	return req, res, ok
}

type AbsorbCubeReq struct {
	CubeID                     uint    `json:"cube_id" binding:"required,gte=1"`
	MemoryGroup                string  `json:"memory_group" binding:"required,max=64"`
	Content                    string  `json:"content" binding:"required"`
	ChunkSize                  int     `json:"chunk_size" binding:"gte=25"`
	ChunkOverlap               int     `json:"chunk_overlap" binding:"gte=0"`
	ChatModelID                uint    `json:"chat_model_id" binding:"required,gte=1"`
	Stream                     bool    `json:"stream" binding:""`
	IsEn                       bool    `json:"is_en"`                                                   // true=English, false=Japanese (default)
	HalfLifeDays               float64 `json:"half_life_days" binding:"omitempty,gte=1"`                // 価値が半減する日数 (デフォルト: 30)
	PruneThreshold             float64 `json:"prune_threshold" binding:"omitempty,gte=0,lte=1"`         // 削除対象となるThickness閾値 (デフォルト: 0.1)
	MinSurvivalProtectionHours float64 `json:"min_survival_protection_hours" binding:"omitempty,gte=0"` // 新規知識の最低生存保護期間 (デフォルト: 72時間)
	MdlKNeighbors              int     `json:"mdl_k_neighbors" binding:"omitempty,gte=1"`               // MDL判定時の近傍ノード数 (デフォルト: 5)
}

func AbsorbCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (AbsorbCubeReq, rtres.AbsorbCubeRes, bool) {
	ok := true
	req := AbsorbCubeReq{}
	res := rtres.AbsorbCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindJSON(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type ExportCubeReq struct {
	CubeID uint `form:"cube_id" binding:"required,gte=1"`
}

func ExportCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (ExportCubeReq, rtres.ExportCubeRes, bool) {
	ok := true
	req := ExportCubeReq{}
	res := rtres.ExportCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindQuery(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type GenKeyCubeReq struct {
	Permissions model.CubePermissions `json:"permissions"`
	ExpireAt    *string               `json:"expire_at"` // ISO8601 or YYYY-MM-DD...
}

// GenKeyCubeReqBind binds multipart form data for GenKey API
// - file: .cube file (handled in BL)
// - permissions: JSON string form field
// - expire_at: optional string form field
func GenKeyCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (GenKeyCubeReq, rtres.GenKeyCubeRes, bool) {
	req := GenKeyCubeReq{}
	res := rtres.GenKeyCubeRes{Errors: []rtres.Err{}}
	// Parse permissions from form field as JSON string
	permStr := c.PostForm("permissions")
	if permStr != "" {
		if err := json.Unmarshal([]byte(permStr), &req.Permissions); err != nil {
			res.Errors = append(res.Errors, rtres.Err{Field: "permissions", Message: "Invalid JSON format"})
			return req, res, false
		}
	} else {
		res.Errors = append(res.Errors, rtres.Err{Field: "permissions", Message: "Required field"})
		return req, res, false
	}
	// Parse expire_at from form field
	expireStr := c.PostForm("expire_at")
	if expireStr != "" {
		req.ExpireAt = &expireStr
	}

	return req, res, true
}

type ImportCubeReq struct {
	Key             string `json:"key" binding:"required"`
	Name            string `json:"name" binding:"required,max=50"`
	Description     string `json:"description" binding:"required,max=255"`
	EmbeddingApiKey string `json:"embedding_api_key" binding:"required"`
}

// ImportCubeReqBind binds multipart form data for Import API
// - file: .cube file
// - key: key string
// - name: new cube name
// - description: optional description
// - embedding_api_key: embedding API key
func ImportCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (ImportCubeReq, rtres.ImportCubeRes, bool) {
	req := ImportCubeReq{}
	res := rtres.ImportCubeRes{Errors: []rtres.Err{}}
	// Parse key from form field
	keyStr := c.PostForm("key")
	if keyStr == "" {
		res.Errors = append(res.Errors, rtres.Err{Field: "key", Message: "Required field."})
		return req, res, false
	}
	req.Key = keyStr
	// Parse name from form field
	nameStr := c.PostForm("name")
	if nameStr == "" {
		res.Errors = append(res.Errors, rtres.Err{Field: "name", Message: "Required field."})
		return req, res, false
	}
	req.Name = nameStr
	// Parse description from form field
	descStr := c.PostForm("description")
	if descStr == "" {
		res.Errors = append(res.Errors, rtres.Err{Field: "description", Message: "Required field."})
		return req, res, false
	}
	req.Description = descStr
	// Parse embedding_api_key from form field
	embKeyStr := c.PostForm("embedding_api_key")
	if embKeyStr == "" {
		res.Errors = append(res.Errors, rtres.Err{Field: "embedding_api_key", Message: "Required field."})
		return req, res, false
	}
	req.EmbeddingApiKey = embKeyStr
	return req, res, true
}

type ReKeyCubeReq struct {
	CubeID uint   `json:"cube_id" binding:"required,gte=1"`
	Key    string `json:"key" binding:"required"`
}

func ReKeyCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (ReKeyCubeReq, rtres.ReKeyCubeRes, bool) {
	ok := true
	req := ReKeyCubeReq{}
	res := rtres.ReKeyCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindJSON(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type QueryCubeReq struct {
	CubeID                  uint    `json:"cube_id" binding:"required,gte=1"`
	MemoryGroup             string  `json:"memory_group" binding:"required,max=64"`
	Text                    string  `json:"text" binding:"required"`
	Type                    uint8   `json:"type" binding:"required,gte=1,lte=11"`                      // 検索タイプ
	SummaryTopk             int     `json:"summary_topk" binding:"omitempty,gte=0"`                    // 要約文の上位k件を取得
	ChunkTopk               int     `json:"chunk_topk" binding:"omitempty,gte=0"`                      // チャンクの上位k件を取得
	EntityTopk              int     `json:"entity_topk" binding:"omitempty,gte=0"`                     // エンティティの上位k件を対象にグラフを取得
	FtsType                 uint8   `json:"fts_type" binding:"omitempty,gte=0,lte=2"`                  // FTSレイヤー: 0=nouns, 1=nouns_verbs, 2=all
	FtsTopk                 int     `json:"fts_topk" binding:"omitempty,gte=0"`                        // FTS拡張Top-K (0=disabled)
	ThicknessThreshold      float64 `json:"thickness_threshold" binding:"omitempty,gte=0,lte=1"`       // エッジ足切り閾値 (デフォルト: 0.3)
	ConflictResolutionStage uint8   `json:"conflict_resolution_stage" binding:"omitempty,gte=0,lte=2"` // 矛盾解決ステージ: 0=なし, 1=Stage1のみ, 2=Stage1+2
	ChatModelID             uint    `json:"chat_model_id" binding:"required,gte=1"`
	Stream                  bool    `json:"stream" binding:""`
	IsEn                    bool    `json:"is_en"` // true=English, false=Japanese (default)
}

func QueryCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (QueryCubeReq, rtres.QueryCubeRes, bool) {
	ok := true
	req := QueryCubeReq{}
	res := rtres.QueryCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindJSON(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type MemifyCubeReq struct {
	CubeID                  uint   `json:"cube_id" binding:"required,gte=1"`
	MemoryGroup             string `json:"memory_group" binding:"required,max=64"`
	Epochs                  int    `json:"epochs" binding:"omitempty,gte=0"`
	PrioritizeUnknowns      bool   `json:"prioritize_unknowns" binding:"boolean"`
	ConflictResolutionStage uint8  `json:"conflict_resolution_stage"` // 0=none, 1=stage1, 2=stage1+2
	ChatModelID             uint   `json:"chat_model_id" binding:"required,gte=1"`
	Stream                  bool   `json:"stream" binding:""`
	IsEn                    bool   `json:"is_en"` // true=English, false=Japanese (default)
}

func MemifyCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (MemifyCubeReq, rtres.MemifyCubeRes, bool) {
	ok := true
	req := MemifyCubeReq{}
	res := rtres.MemifyCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindJSON(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

type DeleteCubeReq struct {
	CubeID uint `form:"cube_id" binding:"required,gte=1"`
}

func DeleteCubeReqBind(c *gin.Context, u *rtutil.RtUtil) (DeleteCubeReq, rtres.DeleteCubeRes, bool) {
	ok := true
	req := DeleteCubeReq{}
	res := rtres.DeleteCubeRes{Errors: []rtres.Err{}}
	if err := c.ShouldBindQuery(&req); err != nil {
		res.Errors = u.GetValidationErrs(err)
		ok = false
	}
	return req, res, ok
}

File: src-from/mode/rt/rtreq/cubes_req.rs
use serde::Deserialize;
use garde::Validate;
use utoipa::{IntoParams, ToSchema};
use crate::mode::rt::rterr::rterr::*;

// ============================================================
// Search
// ============================================================
#[derive(Deserialize, IntoParams, Validate, ToSchema)]
pub struct SearchCubesReq {
    #[schema(example = "Research")]
    #[garde(custom(length_simple_err(0, 50)))]
    pub name: Option<String>,

    #[schema(example = "Description search")]
    #[garde(custom(length_simple_err(0, 255)))]
    pub description: Option<String>,

    #[schema(default = 25, example = 25)]
    #[garde(custom(range_err(Some(1u16), Some(25u16))))]
    pub limit: Option<u16>,

    #[schema(default = 0, example = 0)]
    #[garde(custom(range_err(Some(0u16), None)))]
    pub offset: Option<u16>,
}

// ============================================================
// Create
// ============================================================
#[derive(Deserialize, Validate, ToSchema)]
pub struct CreateCubeReq {
    #[schema(example = "Physics Research")]
    #[serde(default)]
    #[garde(custom(required_simple_err(1, 50)))]
    pub name: String,

    #[schema(example = "Cube for physics papers")]
    #[garde(custom(length_simple_err(0, 255)))]
    pub description: Option<String>,

    #[schema(example = "openai")]
    #[serde(default)]
    #[garde(custom(required_simple_err(1, 50)))]
    pub embedding_provider: String,

    #[schema(example = "text-embedding-3-small")]
    #[serde(default)]
    #[garde(custom(required_simple_err(1, 100)))]
    pub embedding_model: String,

    #[schema(example = 1536)]
    #[serde(default)]
    #[garde(custom(range_err(Some(1i32), None)))]
    pub embedding_dimension: i32,

    #[schema(example = "sk-proj-...")]
    #[serde(default)]
    #[garde(custom(required_simple_err(1, 1024)))]
    pub embedding_api_key: String,

    #[schema(example = "https://api.openai.com/v1")]
    #[garde(custom(length_simple_err(0, 255)))]
    pub embedding_base_url: Option<String>,
}

// ============================================================
// Absorb
// ============================================================
#[derive(Deserialize, Validate, ToSchema)]
pub struct AbsorbCubeReq {
    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub cube_id: u32,

    #[schema(example = "group1")]
    #[garde(custom(length_simple_err(0, 64)))]
    pub memory_group: Option<String>,

    #[schema(example = "Content text...")]
    #[garde(custom(length_simple_err(0, 1000000)))] // Large limit for content
    pub content: Option<String>,

    #[schema(example = 512)]
    #[garde(custom(range_err(Some(25i32), None)))]
    pub chunk_size: Option<i32>,

    #[schema(example = 16)]
    #[garde(custom(range_err(Some(0i32), None)))]
    pub chunk_overlap: Option<i32>,

    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub chat_model_id: Option<u32>,

    #[schema(example = false)]
    #[garde(skip)]
    pub stream: Option<bool>,

    #[schema(example = false)]
    #[garde(skip)]
    pub as_json: Option<bool>,

    #[schema(example = false)]
    #[garde(skip)]
    pub is_en: Option<bool>,

    #[schema(example = 30.0)]
    #[garde(custom(range_err(Some(1.0f64), None)))]
    pub half_life_days: Option<f64>,

    #[schema(example = 0.1)]
    #[garde(custom(range_err(Some(0.0f64), Some(1.0f64))))]
    pub prune_threshold: Option<f64>,

    #[schema(example = 72.0)]
    #[garde(custom(range_err(Some(0.0f64), None)))]
    pub min_survival_protection_hours: Option<f64>,

    #[schema(example = 5)]
    #[garde(custom(range_err(Some(1i32), None)))]
    pub mdl_k_neighbors: Option<i32>,
}

// ============================================================
// Query
// ============================================================
#[derive(Deserialize, Validate, ToSchema)]
pub struct QueryCubeReq {
    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub cube_id: u32,

    #[schema(example = "legal_expert")]
    #[garde(custom(length_simple_err(0, 64)))]
    pub memory_group: Option<String>,

    #[schema(example = "Question?")]
    #[garde(custom(length_simple_err(0, 10000)))]
    pub text: Option<String>,

    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u8), Some(11u8))))]
    pub r#type: Option<u8>,

    #[schema(example = 3)]
    #[garde(custom(range_err(Some(0i32), None)))]
    pub summary_topk: Option<i32>,

    #[schema(example = 3)]
    #[garde(custom(range_err(Some(0i32), None)))]
    pub chunk_topk: Option<i32>,

    #[schema(example = 3)]
    #[garde(custom(range_err(Some(0i32), None)))]
    pub entity_topk: Option<i32>,

    #[schema(example = 0)]
    #[garde(custom(range_err(Some(0u8), Some(2u8))))]
    pub fts_type: Option<u8>,

    #[schema(example = 0)]
    #[garde(custom(range_err(Some(0i32), None)))]
    pub fts_topk: Option<i32>,

    #[schema(example = 0.3)]
    #[garde(custom(range_err(Some(0.0f64), Some(1.0f64))))]
    pub thickness_threshold: Option<f64>,

    #[schema(example = 2)]
    #[garde(custom(range_err(Some(0u8), Some(2u8))))]
    pub conflict_resolution_stage: Option<u8>,

    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub chat_model_id: Option<u32>,

    #[schema(example = false)]
    #[garde(skip)]
    pub stream: Option<bool>,

    #[schema(example = false)]
    #[garde(skip)]
    pub as_json: Option<bool>,

    #[schema(example = false)]
    #[garde(skip)]
    pub is_en: Option<bool>,
}

// ============================================================
// Memify
// ============================================================
#[derive(Deserialize, Validate, ToSchema)]
pub struct MemifyCubeReq {
    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub cube_id: u32,

    #[schema(example = "group1")]
    #[garde(custom(length_simple_err(0, 64)))]
    pub memory_group: Option<String>,

    #[schema(example = 10)]
    #[garde(custom(range_err(Some(0i32), None)))]
    pub epochs: Option<i32>,

    #[schema(example = false)]
    #[garde(skip)]
    pub prioritize_unknowns: Option<bool>,

    #[schema(example = 2)]
    #[garde(custom(range_err(Some(0u8), Some(2u8))))]
    pub conflict_resolution_stage: Option<u8>,

    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub chat_model_id: Option<u32>,

    #[schema(example = false)]
    #[garde(skip)]
    pub stream: Option<bool>,

    #[schema(example = false)]
    #[garde(skip)]
    pub as_json: Option<bool>,

    #[schema(example = false)]
    #[garde(skip)]
    pub is_en: Option<bool>,
}

// ============================================================
// GenKey
// ============================================================
// In Go this is Multipart with 'permissions' JSON string.
// In Rust strict rules, we use Request structs. Only Absorb/Import use multipart for files usually.
// But GenKey requires .cube upload? Go: `file formData file true`.
// Wait, `GenKeyCube` in Go takes a FILE? 
// "Exportされた.cubeファイルをアップロードして鍵を発行" -> Yes.
// For now, we define the Struct part (permissions).
#[derive(Deserialize, Validate, ToSchema)]
pub struct GenKeyCubeReq {
    #[schema(example = "{}")]
    #[garde(skip)]
    pub permissions: String, // JSON string

    #[schema(example = "2026-12-31T23:59:59")]
    #[garde(skip)] // Datetime validation needed? Go uses ISO8601
    pub expire_at: Option<String>,
}

// ============================================================
// ReKey
// ============================================================
#[derive(Deserialize, Validate, ToSchema)]
pub struct ReKeyCubeReq {
    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub cube_id: u32,

    #[schema(example = "new-secret-key...")]
    #[garde(custom(length_simple_err(1, 1024)))]
    pub key: String,
}

// ============================================================
// Import
// ============================================================
// Go: Multipart form data.
#[derive(Deserialize, Validate, ToSchema)]
pub struct ImportCubeReq {
    #[schema(example = "key...")]
    #[garde(custom(length_simple_err(1, 1024)))]
    pub key: String,

    #[schema(example = "My Imported Cube")]
    #[garde(custom(length_simple_err(0, 50)))]
    pub name: Option<String>,

    #[schema(example = "Description...")]
    #[garde(custom(length_simple_err(0, 255)))]
    pub description: Option<String>,

    #[schema(example = "sk-proj-...")]
    #[garde(custom(length_simple_err(0, 1024)))]
    pub embedding_api_key: Option<String>,
}

// ============================================================
// Delete (Query)
// ============================================================
#[derive(Deserialize, Validate, ToSchema, IntoParams)]
pub struct DeleteCubeReq {
    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub cube_id: u32,
}

// ============================================================
// Export (Query)
// ============================================================
#[derive(Deserialize, Validate, ToSchema, IntoParams)]
pub struct ExportCubeReq {
    #[schema(example = 1)]
    #[garde(custom(range_err(Some(1u32), None)))]
    pub cube_id: u32,
}
